### PR TITLE
Fix #11761: Functional Induction throws unrecoverable error.

### LINKED ISSUE #11761: Functional Induction throws unrecoverable error
#### Description of the problem

The tactic `functional induction` can throw a typing error that is not recoverable within Ltac, i.e. it cannot be caught by `try`. Example:
```coq
Require Import FunInd.

Function f (n: nat):= n.

Goal True.
Fail try functional induction f True.
```
This is due to the peculiar syntax of the functional induction. Normally, this would be written as `functional induction (f True)`, but the tactic allows the parentheses to be left out. The problem occurs here:
https://github.com/coq/coq/blob/8fb4fdf4eb0e817412b749794494219597ce3153/plugins/funind/g_indfun.mlg#L98-L107
On line 104 a new term is created that is not type-checked. Eventually, the term ends up here:
https://github.com/coq/coq/blob/8fb4fdf4eb0e817412b749794494219597ce3153/tactics/tactics.ml#L4667
There the type is checked and an exception is thrown, which is of course not catchable within Ltac.
I would submit a pull request that adds the type checking but I'm not entirely sure which function is best for this...

#### Coq Version

The Coq Proof Assistant, version 8.11.0 (January 2020)

### PR DESCRIPTION
**Pull request by Matafou**
**Kind:** documentation / bug fix / feature / performance / infrastructure.

Fixes / closes #11761

### CODE REVIEW

**Comment by Matafou** [comment]
@ejgallego is there a way to simplify this? Ideally the grammar rule should take a list of constr_expr and we should carry on the whole elaboration on the application of all to the first.
Instead in this PR I take the constr and check that applying them is OK, but 1) it forces me to open the goal 2) it doe not support implicit arguments.

File: plugins/funind/g_indfun.mlg [hunk 1]
```diff
@@ -97,13 +97,25 @@ let functional_induction b c x pat =
...
+          (fun x -> functional_induction true c x pat) princl
+     | c::cl ->
+        let appl = EConstr.applist(c,cl) in
+        let check =
+          Proofview.Goal.enter begin fun gl ->
+            let env = Tacmach.New.pf_env gl in
+            let sigma = Tacmach.New.project gl in
+            let _,_ = Typing.type_of env sigma appl in
+            Tacticals.New.tclIDTAC
+            end in
```
Comment (Matafou) [inline plugins/funind/g_indfun.mlg:114]: this build an idtac that performs the type checking. There must be a cleaner way...

**Comment by JasonGross** [comment]
What if you take `uconstr` rather than `constr`?  Or why not just take a single `constr` (at a level that makes you not need parentheses), and let the extreme machinery do the type-checking and application for you?

As for typing, you can probably use tclGETENV and project sigma from that?  Or you can look at how the `type of` tactic argument is implemented?

**Comment by Matafou** [comment]
If you put parenthesis around `(f True)`, then this is one term and the usual elaboration is done. But without parenthesis this is n different independent terms. I don't know how to act "as if" there were parenthesis.

With `uconst` is it apply the first to the list of others and have elaboration?

**Comment by Matafou** [comment]
Sorry I misread your message. How can I have one constr without parenthesis here?

**Comment by JasonGross** [comment]
> Sorry I misread your message. How can I have one constr without parenthesis here?

I think we need @coq/parsing-maintainers or @ppedrot here, since I don't know the `TACTIC EXTEND` machinery well enough.  How do we get a `constr` argument that lives at level 200 (or similar)?

**Comment by JasonGross** [comment]
Note that the way this is done for built-in vernaculars is as in https://github.com/coq/coq/blob/cc54af3842cbf99f169f7937b0e31f737652bd3a/vernac/g_vernac.mlg#L676-L681

**Comment by JasonGross** [comment]
Ah, according to https://github.com/coq/coq/issues/12302#issuecomment-627849987, if you use `lconstr` rather than `constr`, you won't need parentheses to pass an application.

**Comment by ejgallego** [comment]
> Ah, according to [#12302 (comment)](https://github.com/coq/coq/issues/12302#issuecomment-627849987), if you use `lconstr` rather than `constr`, you won't need parentheses to pass an application.

I think that's correct.

**Comment by Matafou** [comment]
Indeed, very nice. Thanks!

**Comment by Matafou** [comment]
Let us see if something breaks. It may be a very rare source of incompatibility.

**Comment by Matafou** [comment]
I think the test failed for other reasons but as it is the test suite will fail. Since now
```functional induction List.app A l l'.``
will fail now because the implicit A should be omitted.
I think it is better like this and hence worth the incompatibility. It is easy to fix and make it pass both versions by putting parenthesis.
How can relaunch the test? So that we see how many things break. Do I need to fix the test suite first?

**Comment by ejgallego** [comment]
I relaunched the test manually, let's see what the impact is.

**Comment by Matafou** [comment]
Thanks.
If it goes merged I will first also adapt "soft functional induction" the same way.
Also should I provide something in doc/changelog if it is merged?

**Comment by ejgallego** [comment]
> Also should I provide something in doc/changelog if it is merged?

Indeed, I think this need a changelog entry.

**Comment by ejgallego** [comment]
> If it goes merged I will first also adapt "soft functional induction" the same way.

I'd suggest you do that in a separate PR as to keep possible compat problems separate.

**Comment by ejgallego** [comment]
Compatibility seems good, we can try to consider this fix for the beta, @Matafou please fix the test-suite case.

**Comment by Matafou** [comment]
I don't know why gitlab CI pipeline failed.
What does it mean when it is tagged "failed" but no job was triggered?

**Comment by JasonGross** [comment]
There should be a little "error" at the top that when you hover over it, it tells you about some rate limit error.  It means that there were already 500(?) jobs running, and GitLab refuses to auto-trigger more jobs at that point.  (We have an issue tracking this, but I'm not going to bother to look it up right now.  Let me know if you can't find it.). The solution is to just click "retry" manually (you need to be logged in to GitLab and might need to force-refresh the page), as there's no limit on manually triggered jobs.

**Comment by Matafou** [comment]
Thanks, I see the hover popup, but even after logging in gitlab and refreshing I don't see the "retry" button. Maybe I need to be in a gitlab team?

**Comment by Zimmi48** [comment]
Yes, you can request to join by going to the organization page.

**Comment by ejgallego** [comment]
Rebased to see if the CI is fixed, but looks like a problem with the changelog entry.

**Comment by Zimmi48** [comment]
I'll take care of fixing the changelog entry!

**Comment by Zimmi48** [comment]
Done!

**Review by ejgallego — APPROVED** [review]

**Pull request merged**